src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -259,417 +258,482 @@ AND NOT EXISTS (
+                "different owner: {}, {}", ownerId, consumer);
 
-            if (ownerId != null && !ownerId.equals(consumer.getOwnerId())) {
-                // Both a consumer and an owner were specified, but the consumer belongs to a different owner.
-                // We can't possibly match a pool on two owners, so we can just abort immediately with an
-                // empty page
-                log.warn("Attempting to filter entitlement pools by owner and a consumer belonging to a " +
-                    "different owner: {}, {}", ownerId, consumer);
+            return emptyPage();
+        }
 
-                return emptyPage();
-            }
+        CriteriaBuilder builder = this.getEntityManager().getCriteriaBuilder();
+        CriteriaQuery<Pool> query = builder.createQuery(Pool.class);
+        Root<Pool> root = query.from(Pool.class);
+
+        Join<Pool, Product> product = root.join(Pool_.product);
+
+        List<Predicate> predicates = new ArrayList<>();
+        Predicate securityPredicate = this.getSecurityPredicate(Pool.class, builder, root);
+        if (securityPredicate != null) {
+            predicates.add(securityPredicate);
+        }
 
-            // We'll set the owner restriction later
+        if (consumer != null) {
             ownerId = consumer.getOwnerId();
+        }
 
-            ConsumerType ctype = this.consumerTypeCurator.getConsumerType(consumer);
-            if (ctype.isManifest()) {
-                DetachedCriteria hostPoolSubquery = DetachedCriteria.forClass(Pool.class, "PoolI")
-                    .createAlias("PoolI.attributes", "attrib")
-                    .setProjection(Projections.id())
-                    .add(Property.forName("Pool.id").eqProperty("PoolI.id"))
-                    .add(Restrictions.eq("attrib.indices", Pool.Attributes.REQUIRES_HOST));
+        Predicate consumerPredicate = getConsumerPredicate(query, root, builder, consumer, ownerId);
+        if (consumerPredicate != null) {
+            predicates.add(consumerPredicate);
+        }
 
-                criteria.add(Subqueries.notExists(hostPoolSubquery));
-            }
-            else if (!consumer.isGuest()) {
-                criteria.add(Restrictions.not(
-                    this.addAttributeFilterSubquery(Pool.Attributes.VIRT_ONLY, Arrays.asList("true"))));
-            }
-            else if (consumer.hasFact(Consumer.Facts.VIRT_UUID)) {
-                String uuidFact = consumer.getFact(Consumer.Facts.VIRT_UUID);
-                Consumer host = null;
+        if (ownerId != null) {
+            predicates.add(builder.equal(root.get(Pool_.owner).get(Owner_.id), ownerId));
+        }
 
-                if (uuidFact != null) {
-                    host = this.consumerCurator.getHost(uuidFact, ownerId);
-                }
+        List<Predicate> poolFilterPredicates = getQualifierPredicates(query, root, builder, product,
+            qualifier);
+        predicates.addAll(poolFilterPredicates);
```

Comment:
Stylistic comment: you can probably compress these to one line by getting rid of the intermediary variable:

`predicates.addAll(this.getQualifierPredicates(...));`

Reply:
This was updated to use the Optional return type